### PR TITLE
REF: Unify _set_noconvert_dtype_columns for parsers

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #39345
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

I unified the code which led to a bugfix. The test failed for the python parser case previously

### CODE REVIEW

File: pandas/io/parsers/base_parser.py [hunk 1]
```diff
@@ -546,6 +548,65 @@ def _convert_to_ndarrays(
                 print(f"Filled {na_count} NA values in column {c!s}")
         return result
 
+    def _set_noconvert_dtype_columns(self, col_indices, names):
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:551]: can you type this in any way? (esp the return value) and add a doc-string

File: pandas/io/parsers/base_parser.py [hunk 2]
```diff
@@ -546,6 +548,65 @@ def _convert_to_ndarrays(
...
+        Set the columns that should not undergo dtype conversions.
+
+        Currently, any column that is involved with date parsing will not
+        undergo such conversions.
+        """
+        noconvert_columns = set()
+        if self.usecols_dtype == "integer":
+            # A set of integers will be converted to a list in
+            # the correct order every single time.
+            usecols = list(self.usecols)
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:562]: -> sorted

File: pandas/io/parsers/base_parser.py [hunk 3]
```diff
@@ -546,6 +548,65 @@ def _convert_to_ndarrays(
...
+            # "List[Any]")  [assignment]
+            usecols = None  # type: ignore[assignment]
+
+        def _set(x):
+            if usecols is not None and is_integer(x):
+                x = usecols[x]
+
+            if not is_integer(x):
+                x = col_indices[names.index(x)]
+
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:582]: can you just return x and do the `.add()` in the caller, its much more obvious.

also type this function if you can

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers/base_parser.py (hunk 1 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:551]: Done

File: pandas/io/parsers/base_parser.py (hunk 3 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py:582]: Typing is a PITA. Mypy does not recognize the isinteger check. Could do with assert but would make this more complicated

File: pandas/io/parsers/base_parser.py [hunk 4]
```diff
@@ -546,6 +548,77 @@ def _convert_to_ndarrays(
...
+        elif callable(self.usecols) or self.usecols_dtype not in ("empty", None):
+            # The names attribute should have the correct columns
+            # in the proper order for indexing with parse_dates.
+            usecols = col_indices
+        else:
+            # Usecols is empty.
+
+            # pandas\io\parsers.py:2030: error: Incompatible types in
+            # assignment (expression has type "None", variable has type
+            # "List[Any]")  [assignment]
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:585]: i think if you predeclare at the top
```
usecols: Optional[List[Any]] = []
```
you can remove the ignore

File: pandas/io/parsers/base_parser.py (hunk 3 above)
Comment (jreback) [reply pandas/io/parsers/base_parser.py:594]: kk

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. if you can try that suggestion and if it works ping (or comment that its a nogo)

File: pandas/tests/io/parser/test_parse_dates.py [hunk 5]
```diff
@@ -1603,3 +1603,21 @@ def test_date_parser_and_names(all_parsers):
...
+    2,4,20-09-01-01
+    """
+
+    parser = all_parsers
+    result = parser.read_csv(
+        StringIO(data),
+        parse_dates=[1],
+        usecols=[1, 2],
+        thousands="-",
+    )
```
Comment (jreback) [inline pandas/tests/io/parser/test_parse_dates.py:1621]: this is a bug yes? can you add a whatsnew note

Comment (phofl) [reply pandas/tests/io/parser/test_parse_dates.py:1621]: Yep, added

File: pandas/io/parsers/base_parser.py (hunk 4 above)
Comment (phofl) [reply pandas/io/parsers/base_parser.py]: Yep, this works, typed it a bit more specific.
I've got #39342 for the mypy errors in there, will rebase when this is merged

**Comment by phofl** [comment]
@jreback green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks !

**Pull request merged**